### PR TITLE
Windows RBD fixes

### PR DESCRIPTION
**Pull request by petrutlucian94**
Windows RBD fixes

This PR includes a few fixes targeting ``rbd-wnbd``:

* New rbd mappings are now started by the ceph service, thus no longer being tied to the user session
* Fixed the block count overflow for image larger than 2TB
* Added a “--non-persistent” map option
* Limited the scope of the commands executed by the Ceph service
* Added configurable service start and image map timeouts. Also fixed an issue where the service would hang indefinitely when failing to remap images
* Changed the way in which the service handles remap failures. Previously, it was getting marked as stopped but already started daemons kept running. Now, by default, the service will keep running unless the “--remap-failure-fatal” option is set, in which case it will cleanup existing mappings before shutting down. 
* Fixed a few error propagation issues
* Moved debug symbols to separate files

Issue: https://tracker.ceph.com/issues/49102

Signed-off-by: Lucian Petrut lpetrut@cloudbasesolutions.com

### CODE REVIEW

**Comment by petrutlucian94** [comment]
Tracked by https://tracker.ceph.com/issues/49102

File: src/tools/rbd_wnbd/rbd_wnbd.cc [hunk 1]
```diff
@@ -540,9 +573,156 @@ class RBDService : public ServiceBase {
...
+      // accept a new connection.
+      HANDLE handler_thread = CreateThread(
+        NULL, 0, (LPTHREAD_START_ROUTINE) handle_connection, pipe_handle, 0, 0);
+      if (!handler_thread) {
+        err = GetLastError();
+        derr << "Could not start pipe connection handler thread: "
+             << win32_strerror(err) << dendl;
+        CloseHandle(pipe_handle);
+      }
+      else {
```
Comment (dillaman) [inline src/tools/rbd_wnbd/rbd_wnbd.cc:846]: Nit: combine `} else {`

File: src/tools/rbd_wnbd/rbd_wnbd.cc [hunk 2]
```diff
@@ -540,9 +573,156 @@ class RBDService : public ServiceBase {
...
+    int create_pipe_server() {
+      HANDLE handler_thread = CreateThread(
+        NULL, 0, (LPTHREAD_START_ROUTINE) pipe_server_loop, NULL, 0, 0);
+      DWORD err = 0;
+
+      if (!handler_thread) {
+        err = GetLastError();
+        derr << "Could not start pipe server: " << win32_strerror(err) << dendl;
+      }
+      else {
```
Comment (dillaman) [inline src/tools/rbd_wnbd/rbd_wnbd.cc:874]: Nit: combine to one line

File: src/common/win32/service.cc [hunk 3]
```diff
@@ -74,22 +74,30 @@ void WINAPI ServiceBase::run()
...
   int err = shutdown_hook();
   if (err) {
     derr << "Shutdown service hook failed. Error code: " << err << dendl;
-    set_status(original_state);
+    if (ignore_errors) {
+      derr << "Ignoring shutdown hook failure, marking the service as stopped."
+           << dendl;
+      set_status(SERVICE_STOPPED);
+    }
+    else {
```
Comment (dillaman) [inline src/common/win32/service.cc:97]: Nit: combine lines

File: src/tools/rbd_wnbd/rbd_wnbd.cc [hunk 4]
```diff
@@ -879,8 +882,15 @@ class RBDService : public ServiceBase {
...
       int r = restart_registered_mappings(
         thread_count, service_start_timeout, image_map_timeout);
-      if (r)
-        return r;
+      if (r) {
+        if (remap_failure_fatal) {
+          derr << "Couldn't remap all images. Cleaning up." << dendl;
+          return r;
+        }
+        else {
```
Comment (dillaman) [inline src/tools/rbd_wnbd/rbd_wnbd.cc:890]: Nit: combine

File: README.windows.rst [hunk 5]
```diff
@@ -181,15 +181,20 @@ within ``ceph.conf`` for the time being.
...
+usually when the host boots.
 
-All mappings are currently persistent, being recreated when the service starts,
-unless explicitly unmapped. The service disconnects the mappings when being
-stopped. This also allows adjusting the Windows service start order so that rbd
-images can be mapped before starting services that may depend on it, such as
-VMMS.
+Note that only one such service may run per host.
+
+By default, all image mappings are persistent. Non persistent mappings can be
```
[the remark this answers is not in the document]
Comment (petrutlucian94) [reply README.windows.rst]: Whops, thanks.

**Review by dillaman — APPROVED** [review]
lgtm

**Pull request merged**